### PR TITLE
Header Stream

### PR DESCRIPTION
**Pull request by fosskers**
This PR adds an endpoint that sends a stream of `BlockHeader`s as Server-Sent Events for consumption by our block explorer.

Can be tested with:
```
chainweb-node --chainweb-version=development --port=8082 --log-level=warn --header-stream

curl -k -v https://localhost:8082/chainweb/0.0/development/header/updates
```

### CODE REVIEW

File: src/Chainweb/CutDB/RestAPI/Server.hs [hunk 1]
```diff
@@ -67,14 +74,27 @@ cutGetHandler db (Just (MaxRank (Max mar))) = liftIO $ do
 cutPutHandler :: CutDb cas -> CutHashes -> Handler NoContent
 cutPutHandler db c = NoContent <$ liftIO (addCutHashes db c)
 
+headerStreamHandler :: CutDb cas -> Tagged Handler Application
+headerStreamHandler db = Tagged $ \req respond -> do
+    chan <- newChan
+    void . SP.mapM_ (writeChan chan . f) . SP.concat $ blockDiffStream db
```
Comment (fosskers) [inline src/Chainweb/CutDB/RestAPI/Server.hs]: `concat` to throw away all the `Left` values (orphaned headers).

File: src/Chainweb/CutDB/RestAPI/Server.hs [hunk 2]
```diff
@@ -67,14 +74,27 @@ cutGetHandler db (Just (MaxRank (Max mar))) = liftIO $ do
...
+    eventSourceAppChan chan req respond
+  where
+    f :: BlockHeader -> ServerEvent
+    f bh = ServerEvent (Just $ fromByteString "BlockHeader") Nothing
+        [ fromLazyByteString . encode . toJSON $ ObjectEncoded bh ]
```
Comment (fosskers) [inline src/Chainweb/CutDB/RestAPI/Server.hs]: Should probably return more than just the header. Needs stats too.

File: src/Chainweb/CutDB/RestAPI.hs [hunk 3]
```diff
@@ -89,3 +94,14 @@ cutApi = Proxy
 someCutApi :: ChainwebVersion -> SomeApi
 someCutApi (FromSing (SChainwebVersion :: Sing v)) = SomeApi $ cutApi @v
 
+-- -------------------------------------------------------------------------- --
+-- BlockHeader Event Stream
+
+type HeaderStreamApi_ = "header" :> "updates" :> Raw
```
Comment (fosskers) [inline src/Chainweb/CutDB/RestAPI.hs]: Probably the wrong endpoint path.

**Comment by larskuhtz** [comment]
I am not sure if this API should be part of the "standard" chainweb-node. I am concerned that we keep adding features and functionality to the consensus nodes eventually overloading them.

HTTP long polling is potentially expensive, since it allows a client to keep a TLS connection along with a Haskell thread open with verify little resource overhead on the client side.

I wonder providing this kind of information should be the role of dedicated chainweb monitoring nodes.

Or, if we really want just a single chainweb-node application that servers everybody's needs (in addition to providing consensus with best possible performance), which should at least make serving certain non-critical APIs optional (via configuration settings)?

**Comment by larskuhtz** [comment]
I'd like use to rather focus on performance and soundness of the core features of a chainweb-node then adding new non-essential APIs and features.

**Comment by fosskers** [comment]
Lars and I have discussed this and I'll be making certain improvements to ensure this endpoint isn't abused.

File: src/Chainweb/BlockHeaderDB/RestAPI/Server.hs [hunk 4]
```diff
@@ -276,3 +295,35 @@ someBlockHeaderDbServer (SomeBlockHeaderDb (db :: BlockHeaderDb_ v c))
...
+headerStreamServer cdb = headerStreamHandler cdb
+
+headerStreamHandler :: forall cas. PayloadCas cas => CutDb cas -> Tagged Handler Application
+headerStreamHandler db = Tagged $ \req respond -> do
+    chan <- newChan
+    void . SP.mapM_ (g >=> writeChan chan . f) . SP.concat $ blockDiffStream db
+    eventSourceAppChan chan req respond
+  where
+    cas :: PayloadDb cas
+    cas = _webBlockPayloadStoreCas $ view cutDbPayloadStore db
```
Comment (fosskers) [inline src/Chainweb/BlockHeaderDB/RestAPI/Server.hs]: Causes a payload DB hit per `BlockHeader`!

File: src/Chainweb/BlockHeaderDB/RestAPI/Server.hs [hunk 5]
```diff
@@ -276,3 +295,35 @@ someBlockHeaderDbServer (SomeBlockHeaderDb (db :: BlockHeaderDb_ v c))
...
+
+    g :: BlockHeader -> IO HeaderUpdate
+    g bh = do
+        x <- casLookupM cas $ _blockPayloadHash bh
+        pure $ HeaderUpdate (ObjectEncoded bh) (length $ _payloadWithOutputsTransactions x)
```
Comment (fosskers) [inline src/Chainweb/BlockHeaderDB/RestAPI/Server.hs]: `Foldable.length` on `Vector` is safe, it's overridden to `V.length` which is `O(1)`.

File: src/Chainweb/Chainweb.hs [hunk 6]
```diff
@@ -265,6 +272,9 @@ pChainwebConfiguration = id
     <*< configCoordinator .:: boolOption_
         % long "mining-coordination"
         <> help "whether to enable external requests for mining work"
+    <*< configHeaderStream .:: boolOption_
+        % long "header-stream"
+        <> help "whether to enable an endpoint for streaming block updates"
```
Comment (fosskers) [inline src/Chainweb/Chainweb.hs]: Off by default. Specifying this flag turns it on.

File: src/Chainweb/RestAPI.hs [hunk 7]
```diff
@@ -210,6 +213,8 @@ someChainwebServer v dbs mr =
         <> someP2pServers v (_chainwebServerPeerDbs dbs)
         <> PactAPI.somePactServers v (_chainwebServerPactDbs dbs)
         <> maybe mempty (Mining.someMiningServer v) mr
+        <> maybe mempty (someHeaderStreamServer v)
+               (bool Nothing (Just ()) hs >> _chainwebServerCutDb dbs)
```
Comment (fosskers) [inline src/Chainweb/RestAPI.hs]: I wish there were better ergonomics between `Bool` and `Maybe`.

File: src/Chainweb/BlockHeaderDB/RestAPI/Server.hs [hunk 8]
```diff
@@ -276,3 +295,35 @@ someBlockHeaderDbServer (SomeBlockHeaderDb (db :: BlockHeaderDb_ v c))
...
+    :: forall cas (v :: ChainwebVersionT)
+    .  PayloadCas cas
+    => CutDb cas
+    -> Server (HeaderStreamApi v)
+headerStreamServer cdb = headerStreamHandler cdb
+
+headerStreamHandler :: forall cas. PayloadCas cas => CutDb cas -> Tagged Handler Application
+headerStreamHandler db = Tagged $ \req respond -> do
+    chan <- newChan
+    void . SP.mapM_ (g >=> writeChan chan . f) . SP.concat $ blockDiffStream db
```
Comment (emilypi) [inline src/Chainweb/BlockHeaderDB/RestAPI/Server.hs]: - What does `Tagged` do here? 

- `void` is unnecessary 

- Why split `g` and `f`? Kleisli fish are nice and all but you've defined the functions locally below, and there doesn't seem to be any multiple usages.

File: src/Chainweb/RestAPI.hs (hunk 7 above)
Comment (emilypi) [reply src/Chainweb/RestAPI.hs]: This is definitely a code smell. `Maybe ()` is isomorphic to `Bool`, so this effectively `bool False True <some bool>`

Comment (fosskers) [reply src/Chainweb/RestAPI.hs]: `_chainwebServerCutDb` gives a `Maybe`, but you're right that I can just put that in place of `Just ()`.

File: src/Chainweb/BlockHeaderDB/RestAPI/Server.hs (hunk 8 above)
Comment (fosskers) [reply src/Chainweb/BlockHeaderDB/RestAPI/Server.hs]: The endpoint that this handler corresponds to ends in `Raw`. This is the strategy that we've discovered for doing Server-Sent Events with Servant.

`void` was indeed necessary.

`f` and `g` are split for cleanliness.

File: src/Chainweb/BlockHeaderDB/RestAPI/Server.hs (hunk 4 above)
Comment (mightybyte) [reply src/Chainweb/BlockHeaderDB/RestAPI/Server.hs]: If this proves to be too much overhead, we can consider removing it in the future.

**Pull request merged**